[Permissions] Clean-up storage permissions and video recording

thomashorta opened the pull request:
Part of #19016
Companion PR: https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/133

Although not directly related to supporting partial media access, this PR cleans up unnecessary storage and media permissions requests. That said, if you use this version on Android 14, giving partial media access will still be broken on flows that need media permissions.

## Notable changes

### Get Video URI from result Intent
Previously we were using `getLastRecordedVideoUri` to get the URI of the video we just requested, instead of relying on the proper way of doing use, which is either setting the `EXTRA_OUTPUT` in the request intent (like we do for images) or using the `data` field from the result `Intent`, which contains the URI for the video, already with proper permissions.

This is also perfect for Android 14 because of the new partial permission grant for videos and images, which makes `getLastRecordedVideoUri` stop working in that scenario since the user would not have given access to the video we just recorded.

Also, this caused the very strange situation of us asking for media permissions when the user selected the Camera as a source, which doesn't make much sense.

Note: Although it should be safe to also remove asking STORAGE permission in older versions, in versions below Android 13 I kept the request for them when asking for the Camera permissions, as it reduces the footprint of changes and is "less dangerous" since there didn't exist such a fragmented storage/media permission model in those versions.

### Fix file downloads and remove unneeded permissions
Same as above, for file downloads, in Android 13 and up we were asking for `READ_MEDIA_*` permissions, which again don't make sense as we are writing to the storage. https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/133 made some changes in the storage permissions and now exposes a specific method for checking `FileDownload` permissions, which for Android Q and up are none if we use `DownloadManager`, as we do.

Other than that, the file download functionality was broken, since it was trying to use a valid URL as file name, which doesn't work because of the special characters present in that string. To fix it I replaced the usage of `guessUrl` with `guessFileName` which provides a valid file name based on the String input.

## Remove unneeded storage permission for media-sharing flows
We support media sharing from other apps via intent filters and we were checking/asking for `READ_MEDIA` or `STORAGE` permissions when apps shared media with us, but in those scenarios, we don't actually need any permissions, as the app providing the media URI already grants access to it. Because of that, I removed the code that asked for those storage permissions.

## Fix the drag-and-drop feature in Aztec
The code implementing the drag-and-drop of media files while in Split Screen was not working correctly, since it was doing some asynchronous processing with a shared array reference but cleaning that array before it could be processed.

To fix it I am making a copy of the array to pass to the async job instead of passing the shared reference.

### Permission clean-ups
Overall clean up code related to permission, media, and Camera. For instance, the deprecated `PhotoPickerActivity` was using `launchCameraForVideo` but that code was never even reached.

## To test
Login to the Jetpack/WordPress app.

### File Download
On Web:
1. Create a post with a File block and upload a PDF (or other non-media file) to it
2. Like that post (on Reader) so you can easily find it on the mobile app Reader

On Jetpack app:
1. In the Reader, find the above post
2. Click the download link (note: currently the Reader shows an extra `Download` text after the file name)
3. (below Android 10) **Verify** storage permission is requested and grant it
4. **Verify** the file is downloaded and opened

Ideally, test this in several Android versions.

### Media Sharing
Make sure Jetpack/WordPress apps DO NOT have any storage nor media permission.

On Gallery/Photos/Files app:
1. Choose a photo/video
2. Share it to Jetpack/WordPress app

On Jetpack/WordPress app:
1. Choose a site and an option (sharing to the media library or to a new post)
2. **Verify** no permission is asked
3. **Verify** the media was properly shared based on the chosen option

### Drag-and-drop media in Aztec Editor
On Web (create a post that forces Aztec on the app):
1. Create a post
2. Go to the Code Editor
3. Add some text to the post body
4. Save as draft

On Jetpack/WordPress app:
1. Put the app in split-screen mode with a gallery/file explorer app
2. Open the above draft post

On Gallery/Photos/Files app:
1. Select a picture/video
2. Drag and drop the media to the Aztec Editor
3. **Verify** no permission is requested
4. **Verify** the media is properly added to the post and uploaded

### Adding Video from camera:
There are several flows to test. Before each of them, make sure to remove all permissions from the app.

1. Create video Story post
   1. (Android 13 and up) **Verify** it doesn't ask for storage/media permissions when requesting the Camera permission
   2. (Below Android 13) **Verify** it still asks for storage permissions when requesting Camera permission
   3. **Verify** it works properly in all versions
2. (Gutenberg) Add Video blocks and choose "Take a Video" as the source
   1. (Android 13 and up) **Verify** it doesn't ask for storage/media permissions when requesting the Camera permission
   2. (Below Android 13) **Verify** it still asks for storage permissions when requesting Camera permission
   3. **Verify** it works properly in all versions
3. (Aztec) Add Video and choose "Take Video" as the source
   1. (Android 13 and up) **Verify** it doesn't ask for storage/media permissions when requesting the Camera permission
   2. (Below Android 13) **Verify** it still asks for storage permissions when requesting Camera permission
   3. **Verify** it works properly in all versions

## Regression Notes
1. Potential unintended areas of impact
Breakage of any flows using Storage or Camera.

7. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

8. What automated tests I added (or what prevented me from doing so)
Updated existing tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
Likely not needed, since there weren't UI changes.

- [ ] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] Large and small screen sizes. (Tablet and smaller phones)
- [x] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

Commit:
Remove usage of getLastRecordedVideoUri and use data from result

When using the ACTION_VIDEO_CAPTURE intent, the Camera returns the video URI in
the data field of the result Intent in `onActivityResult`. This is the preferred
way of retrieving that URI (when EXTRA_OUTPUT is not set in the request intent)
because it already provides the saved file URI with access permissions granted
to the caller application.

This is perfect for Android 14 because of the new partial permission grant for
videos and images, which makes getLastRecordedVideoUri to stop working in that
scenario.

The getLastRecordedVideoUri is being removed from the WordPress-Utils lib to
avoid it being used as well.

Commit:
remove launchCameraForVideo from PhotoPickerAcitivty

This is not being used anymore in the PhotoPickerActivity which doesn't even
check for the response of the ACTION_VIDEO_CAPTURE intent sent by
launchCameraForVideo.

Commit:
Remove duplicate media scan request and clear used Media path

It seems at some point in the past the WPMediaUtils.scanMediaFile was added to
the codebase but the previous way of requesting media scan here was not removed
leading to duplicate media scan request using different methods, which is not
necessary.

Commit:
Remove unused request codes in EditPostSettings onActivityResult

The if condition checked for request codes that were simply ignored inside the
block, so this is just a minor change to improve the code and readability.

Commit:
Copy media list before cleaning saved array

The drag'n'drop feature was not working because this list was cleaned before the
async code was executed. Since that list was passed by reference, it caused the
async code to process an empty list.

Commit:
Use specific checkAndRequestFileDownloadPermission

Commit:
Stop requesting unnecessary storage permissions

When other apps share or drop (in the case of Aztec) images/videos to our app,
the permission to the media from those shared URIs is granted to us, so there's
no need to request for extra storage permissions.

Requesting READ_MEDIA permissions in Android 14 in those cases is even worse if
user has given only partial media permission, as it's not clear what we are
requesting (and not needed anyway).

Commit:
Camera and Storage permissions in WP-Util-Android

Commit:
Replace guessUrl with guessFileName

Not sure if this feature has been broken all this time or something changes in
how that Android method works, but `guessUrl` actually returns a valid URL,
which most of the time is NOT a valid file name, as it contains several `.`, `:`
and `/` characters, so the DownloadManager is not able to download the files.

`guessFileName` on the other hand transforms URLs to valid file names and should
be used in this scenario.

Commit:
Remove unused field in test

Commit:
Remove unused imports in EditorMedia

Commit:
Remove unused imports in ShareIntentReceiverActivity

RenanLukas commented:
Hey @thomashorta , thanks for the PR. I'm sharing the results of my tests.

### File Download
✅ Tested on `Android 11`
✅ Tested on `Android 9`

### Media Sharing
Tested sharing image and video, adding to media library and to new post. 
✅ Tested on `Android 11` 
✅ Tested on `Android 9`

### Drag-and-drop media in Aztec Editor 
✅ Tested on `Android 11`
🟡 Tested on `Android 9`: the dragged image doesn't appear in the post. Since it worked in your emulator, maybe something's wrong with mine. In case someone has a physical device with `9` installed to test this scenario that'd be great.

### Adding Video from camera 
#### Create video Story post
✅ Tested on `Android 13`
🔴  Tested on `Android 11`: hopefully this happened because I was testing on an emulator, but every time I tried to open the camera after granting the camera and storage permissions the app crashed. Besides the fact that it crashed in the emulator, both camera and storage permissions were asked as expected.
```
FATAL EXCEPTION: main
Process: com.jetpack.android.prealpha, PID: 12173
java.lang.IllegalArgumentException: Fail to find supported surface info - CameraId:null
androidx.camera.camera2.impl.Camera2DeviceSurfaceManager.requiresCorrectedAspectRatio(Camera2DeviceSurfaceManager.java:274)
at androidx.camera.core.Preview.updateUseCaseConfig(Preview.java:476)
at androidx.camera.core.UseCase.<init>(UseCase.java:96)
at androidx.camera.core.Preview.<init>(Preview.java:115)
at com.automattic.photoeditor.camera.CameraXBasicHandling.startCamera(CameraXBasicHandling.kt:117)
at com.automattic.photoeditor.camera.CameraXBasicHandling.startUp(CameraXBasicHandling.kt:71)
at com.automattic.photoeditor.camera.CameraXBasicHandling.activate(CameraXBasicHandling.kt:54)
at com.automattic.photoeditor.state.BackgroundSurfaceManager.switchCameraPreviewOn(BackgroundSurfaceManager.kt:204)
at com.wordpress.stories.compose.ComposeLoopFrameActivity.switchCameraPreviewOn(ComposeLoopFrameActivity.kt:1329)
at com.wordpress.stories.compose.ComposeLoopFrameActivity.launchCameraPreview(ComposeLoopFrameActivity.kt:1391)
at com.wordpress.stories.compose.ComposeLoopFrameActivity.access$launchCameraPreview(ComposeLoopFrameActivity.kt:206)
at com.wordpress.stories.compose.ComposeLoopFrameActivity$launchCameraPreviewWithSurfaceSafeguard$1.invokeSuspend(ComposeLoopFrameActivity.kt:1361)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:223)
at android.app.ActivityThread.main(ActivityThread.java:7656)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@bc22a0b, Dispatchers.Main]
```
🔴  Tested on `Android 9`: same crash as `Android 11`, except that the camera opened the first time before starting to crash. Hopefully it's a problem with emulators. 

#### (Gutenberg) Add Video blocks and choose "Take a Video" as the source
✅ Tested on `Android 13` 
✅ Tested on `Android 11`
🔴 Tested on `Android 9`: I'm getting this error when tapping on `"Take video"`. The camera permission was definitely granted.
<img width="300" alt="image" src="https://github.com/wordpress-mobile/WordPress-Android/assets/14964993/bcfa7510-04ee-4974-a419-e4fd5fabccb3">

#### (Aztec) Add Video and choose "Take Video" as the source
✅ Tested on `Android 13` 
✅ Tested on `Android 11`
🔴 Tested on `Android 9`: I'm getting the same error that happened in Gutenberg flow.
<img width="300" alt="image" src="https://github.com/wordpress-mobile/WordPress-Android/assets/14964993/bcfa7510-04ee-4974-a419-e4fd5fabccb3">

irfano commented:
### File Download
🟢 Android 13 🟢 Android 11
🔴 Android 10: I can download the file and then open it manually from the file browser. But it's not opening automatically after tapping the link within the reader, and I get this error `Can't load file (doesn't open)  Display Data [PDF : sample-1-4.pdf] +ContentOpenable, uri: content://com.jetpack.android.beta.provider/external_files/Download/sample-1-4.pdf`
🔴 Android 9: It asked for permission, then downloaded the file. But we are asking for permissions now, and the user can deny and select the "Don't ask again" option. So, we need to handle this case and navigate users to system settings if we're not allowed to show the permission dialog since it's denied.

### Media Sharing
Great catch! ❤️ 
🟢 Android 13 🟢 Android 11 🟢 Android 10 🟢 Android 9

### Drag-and-drop media in Aztec Editor
🟢 Android 13
Note: Drag & drop didn't work on the emulator, so I couldn't test it for other Android versions. 
❓ Do you know why it doesn't work when Aztes is not forced?

### Adding Video from camera
All the cases worked on my 🟢 Android 13 device, but the emulator camera returned errors on previous Android versions, similar to what happened with @RenanLukas. The emulator seems to be the issue.  @thomashorta, were you be able to test them on the emulator?

---
I suggest bumping the milestone to `23.3` then we can have a good time to spend on testing this PR.

thomashorta commented:
Thanks @RenanLukas for the review and test. We already talked offline but in your case, it seems all issues could be emulator-related.

---

Thanks @irfano as well for the review and testing.

> 🔴 Android 10: I can download the file and then open it manually from the file browser. But it's not opening automatically after tapping the link within the reader, and I get this error Can't load file (doesn't open)  Display Data [PDF : sample-1-4.pdf] +ContentOpenable, uri: content://com.jetpack.android.beta.provider/external_files/Download/sample-1-4.pdf

That's odd, but since it works properly in the other versions for you my guess is that something is missing in the emulator base image to properly open this type of file. 

To confirm that, I tested this on a physical device with Android 10 and this is working fine. It uses the Google Drive app for that, which further confirms that theory, as disabling the Drive app shows me a "Can't open file" toast. I believe that the file explorer in the emulator has an embedded PDF reader, but it likely doesn't have a READ intent filter to open files from external sources.

> 🔴 Android 9: It asked for permission, then downloaded the file. But we are asking for permissions now, and the user can deny and select the "Don't ask again" option. So, we need to handle this case and navigate users to system settings if we're not allowed to show the permission dialog since it's denied.

For Android 9 we were already asking for permissions before, this behavior was not changed, so I believe this is an existing bug in our code. Not sure if we should fix it in the scope of this PR though. Wdyt? If you agree, can you open a GH issue with what you saw?

Here is the change in check/request permission on WP-Android:
https://github.com/wordpress-mobile/WordPress-Android/pull/19103/files#diff-f4897a8a09fc8e5a463d003946a0c2d6143f0091caa1bf013c99d3b89a3ac92cR1712

And here's the diff in the Utils lib, which only changed the permission check for Android 10 (Q, or SDK 29) and up, by **removing** the permission request, while **keeping it the same behavior** for any versions below 10:
https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/133/files#diff-b49f41b7026a3874490e0092269a8f784e266a0a1f45272ec850dba17f4f125fR125

> Note: Drag & drop didn't work on the emulator, so I couldn't test it for other Android versions.

So, drag & drop is a super weird feature that I doubt anyone uses. Testing it is a bit complicated too, as not all apps support it in all emulator versions. I was able to test it in the emulator on Android 9 and up using the File Manager. 

To do it you:
1. Open split screen with Jetpack and File Manager
2. Tap and hold an image in the file manager, and release it when it's selected
3. With the image selected you now tap and hold again, **and keep holding for the next step**
4. Drag it to the Jetpack post (Aztec editor)

It's not at all intuitive TBH.

> ❓ Do you know why it doesn't work when Aztec is not forced?

Yep, it looks like this feature was not implemented in Gutenberg. Aztec implements it [here](https://github.com/wordpress-mobile/WordPress-Android/blob/2e6bbb0488debb3b26b14f4e2a383581dd51d3dd/libs/editor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java#L1551-L1615), calling the `onMediaDropped` function [implemented by the editor Activity](https://github.com/wordpress-mobile/WordPress-Android/blob/9b519574b664bd0b2403db37e1e1b871987f363f/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L3269-L3274). I couldn't find any calls for that from Gutenberg.

> @thomashorta, were you be able to test them on the emulator?

Yeah, I was able to test it on the emulator for Android 9, 13, and 14, as well as on physical devices for Android 10 and 13. I got errors when testing on emulators for Android 7 and 8, but I am pretty confident it's related to the emulator Camera implementation.

---

> I suggest bumping the milestone to 23.3 then we can have a good time to spend on testing this PR. 

Great point, I just did it and updated the PT post for it as well since we will definitely be able to merge it today.

irfano commented:
> To confirm that, I tested this on a physical device with Android 10 and this is working fine.

It's great that you could test it. I agree with your thoughts.  _🟢 for the case._

> For Android 9 we were already asking for permissions before, this behavior was not changed, so I believe this is an existing bug in our code. Not sure if we should fix it in the scope of this PR though. Wdyt? If you agree, can you open a GH issue with what you saw?

If it was the same before, I agree it shouldn't be fixed in this PR. I opened an issue for it, https://github.com/wordpress-mobile/WordPress-Android/issues/19131. _🟢 for the case._

> I was able to test it in the emulator on Android 9 and up using the File Manager.

Thanks! I could finally test it on Android 9 and it didn't ask for any permission.  

> Yeah, I was able to test it on the emulator for Android 9, 13, and 14

Good! I agree it should be the emulator issue on my side. I won't try to fix the emulator problem and test on lower APIs since you're already tested. 

All looks fine. I'll do a last round of reviewing the coding side and then approve.

irfano commented on WordPress/src/main/java/org/wordpress/android/util/WPPermissionUtils.java:
```diff
@@ -29,13 +29,17 @@
     // permission request codes - note these are reported to analytics so they shouldn't be changed
+    @Deprecated
+    // This is deprecated to keep to request code taken but media shared to our app does not require permissions
     public static final int SHARE_MEDIA_PERMISSION_REQUEST_CODE = 10;
```
I believe it's safe to remove `SHARE_MEDIA_PERMISSION_REQUEST_CODE` and `EDITOR_DRAG_DROP_PERMISSION_REQUEST_CODE`. Do you know a reason not to remove?

thomashorta replied:
I did add the `@Deprecated` mainly to keep those constant values visibly taken in that context, but you're right, it makes more sense to remove it. I will work on removing it. Thanks!

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -2985,20 +2983,19 @@ private List<Uri> convertStringArrayIntoUrisList(String[] stringArray) {
-                final Intent scanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
-                scanIntent.setData(capturedImageUri);
-                sendBroadcast(scanIntent);
+                mEditorTracker.trackAddMediaFromDevice(mSite, true, false, capturedImageUri);
```
⚠️  **Track issue**: I added an image to the image block by taking a photo. The `editor_photo_added` event was tracked twice. Calling `trackAddMediaFromDevice()` in `addLastTakenPicture` should be redundant.

thomashorta replied:
You are right. It seems that the `OptimizeMediaUseCase` already tracks the media:
https://github.com/wordpress-mobile/WordPress-Android/blob/0002b4ede131f3bcc8f99596098d3832a81ffa95/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeMediaUseCase.kt#L61-L63

This happens both for pictures and videos, so I'm removing the redundant tracking I added for both, and also fixing a bug related to new video tracking, as the code below is trying to check the extension of a media URI instead of a file path (so videos were being tracked as photos):
https://github.com/wordpress-mobile/WordPress-Android/blob/0002b4ede131f3bcc8f99596098d3832a81ffa95/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeMediaUseCase.kt#L47

Commit:
Remove unused permission request codes

Commit:
Fix check for video inside media optimization function

Commit:
Remove duplicate tracking for newly added photos/videos

The `OptimizeMediaUseCase` already tracks photo and video addition for freshly
taken media when we call `EditorMedia#addNewMediaToEditorAsync` functions when
`TAKE_PHOTO` and `TAKE_VIDEO` requests are returned, so no need to call the
tracking manually.

Commit:
Point to Utils lib release version

irfano reviewed (approved):
Thanks for addressing my comments and also fixing the bug you found, @thomashorta. All LGTM! 👍🏻
